tests/force_insert_update/tests.py

```diff
@@ -66,3 +75,153 @@ def test_force_update_on_inherited_model_without_fields(self):
+                Child(id=2, name="Child-2", value=9).save(
+                    force_insert=(SubCounter, Counter)
+                )
+
+        with transaction.atomic():
+            Counter(id=3, name="Grandparent-3", value=10).save()
+            SubCounter(id=3, name="Parent-3", value=10).save()
+            with self.assertRaises(IntegrityError):
+                Child(id=3, name="Child-3", value=10).save(force_insert=(SubCounter,))
+
+        # Backward compatibility tests
+        Counter(id=4, name="Grandparent-4", value=11).save()
+        SubCounter(id=4, name="parent-4", value=11).save()
+        with self.assertNumQueries(3):
+            Child(id=4, name="Child-4", value=11).save(force_insert=True)
+
+        Counter(id=5, name="Grandparent-5", value=11).save()
+        SubCounter(id=5, name="parent-5", value=11).save()
+        with self.assertNumQueries(4):
+            Child(id=5, name="Child-5", value=11).save()
+
+    def test_create_query_count(self):
+        Counter.objects.create(id=1, name="Grandparent-1", value=8)
+        with self.assertNumQueries(2):
+            SubCounter.objects.create(id=1, name="Parent-1", value=8)
+
+        with self.assertNumQueries(3):
+            SubCounter.objects.create(id=2, name="Parent-2", value=8)
+
+    def test_force_insert_in_diamond_inheritance(self):
+        # p, gp DNE
+        with self.assertNumQueries(4):
+            DiamondChild(id=2, name="Child-2", value=8).save(
+                force_insert=(Counter, SubCounter, DiamondParent)
+            )
+
+        with self.assertNumQueries(4):
+            DiamondChild(id=3, name="Child-3", value=8).save(
+                force_insert=(models.Model,)
+            )
+
+        # gp E, p DNE
+        Counter(id=6, name="Grandparent-6", value=8).save()
+        with transaction.atomic():
+            with self.assertRaises(IntegrityError):
+                DiamondChild(id=6, name="Child-6", value=8).save(
+                    force_insert=(models.Model,)
+                )
+
+    def test_validate_force_insert(self):
+        expected_message_1 = (
+            f"Invalid force_insert member. {SubCounter!r} "
+            f"must be a base of {Counter!r}."
+        )
+        with self.assertRaisesMessage(TypeError, expected_message_1):
+            Counter(id=1, name="Parent-1", value=8).save(force_insert=(SubCounter,))
+
+        expected_message_2 = (
+            f"Invalid force_insert member. {TestCase!r} must be a model subclass."
+        )
+
+        with self.assertRaisesMessage(TypeError, expected_message_2):
+            Counter(id=1, name="Parent-1", value=8).save(force_insert=(TestCase,))
+
+        expected_message_3 = "force_insert must be a bool or tuple"
+        with self.assertRaisesMessage(TypeError, expected_message_3):
+            SubCounter(id=1, name="Parent-1", value=8).save(
+                force_insert=[
+                    Counter,
+                ]
+            )
```

Comment:
I decided to reorganize new tests, some of them are redundant, some of them are unnecessary, some of them work without this patch. I moved the last group to a separate PR, check out #17028.